lib/msal-core/src/utils/RequestUtils.ts

```diff
@@ -156,4 +158,24 @@ export class RequestUtils {
         }
         return CryptoUtils.isGuid(correlationId)? correlationId : CryptoUtils.createNewGuid();
     }
+
+    /**
+     * @ignore
+     * 
+     * compute a logical identifier to represent a token request made by a specified account.
+     * if two tuples of <ServerRequestParameters, Account> have the same "uniqueness identifier"
+     * then they can be treated as the same overall request and merged together as an optimization.
+     * @param request 
+     * @param account 
+     */
+    static computeTokenRequestUniquenessIdentifier(request: ServerRequestParameters, account: Account): string {
+        // TODO - resolve exact information that should disambiguate a request
+        let parts = [
+            account.accountIdentifier,
+            request.authority,
+            request.scopes.sort().join(Constants.resourceDelimiter),
+            request.promptValue // a token request to prompt consent should be distinct from one that does not require consent
+        ];
+        return parts.join(Constants.resourceDelimiter).toLowerCase();
+    }
```

Comment:
this function is added to request utils, though it could be somewhere else, and it represents the computation of an identifier for a target request and account that is used to disambiguate the request.

The current implementation of MSAL is essentially the behavior when this function returns only the target scopes from the request, as the library previously only considered scope values as the thing that differentiated different potential requests. However, authority is easily identified as another component that MUST be taken into consideration, or the library will inappropriately return the wrong token to the caller for a different authority, if such an alternate request is ongoing.

Above, I have included a few elements to consider which I suspect to be relevant in considering a request as unique or if it is logically equivalent to some other request that may be ongoing. I'm not sure if things like extra query parameters are enough to warrant a semantic difference and thus be considered in above, but in general, this PR creates the necessary abstraction for you folks to curate the logic over time.

Reply:
We are in the planning/design phase to do work in our libraries (all MSALs, not just MSAL.js) to implement proper throttling mechanisms, and one of the things we will be doing for that is computing a "fingerprint" for each request, and using that to prevent the same request from being made repeatedly and/or unnecessarily. What you have above is actually very close to what has been proposed. 

We'll need to look at this holistically for both implicit and auth code flows (which have some different considerations), and also for the different actions in the library that behavior will control (linking multiple of the "same" concurrent requests together, which is what you have here, and whether or not to make a given request at all, which is the main driver of the throttling initiative). 

We'd love to get your feedback on that as we start building that out. cc: @jmckennon who is leading that for us.

Reply:
Thanks for taking a look so quickly! I haven't worked with typescript before so I was a bit uncertain if I was locating all the necessary areas to change. I actually haven't worked with JavaScript at all before trying-out this MSAL client for a side project; I thought the jokes about supporting IE were just memes but it is actually soul-crushing every time I scroll down to the "supported browsers" section of [some feature](https://developer.mozilla.org/en-US/docs/Web/API/URL/searchParams) or [another](https://developer.mozilla.org/en-US/docs/Web/CSS/Using_CSS_custom_properties) and find everyone EXCEPT IE supports it! haha. The TS compilation of async / await looks to be super useful so I will likely abandon my promise chaining callback hell of a sample for a compiled version ;)

It is interesting to see your team focus / concern on throttling and helping end-user-code not make too many "duplicate" requests in a short window of time. I'm more concerned on the opposite end, what if users make several distinct, concurrent requests, do they all resolve correctly? I think I had previously-mentioned, I actually "don't care so much" whether there is any optimization handled at the MSAL layer, and I would be perfectly fine to have each request resolved on its own without any de-duplication, other than the logic around returning an existing token from the cache if one is available which I obviously would prefer. But I can also appreciate how easy it is to end-up writing such code that falls into redirect-loops and whatnot so having safeguards against that can be helpful.

That bit about the cache is interesting too; I would presume that the logic for finding an existing token in the token cache should accommodate all the necessary considerations for whether a request is duplicate or not, though I guess it wound't take into account "meta-request" properties (like prompt) for such token resolution, which may be necessary for de-duping live requests, so the logic isn't exactly 1:1, but very similar I expect.

I would be interested to see what your request fingerprinting looks like; on what timeline (roughly) are you pursuing those changes?